src/sys/socket/mod.rs

```diff
@@ -2009,12 +2009,116 @@ pub fn socketpair<T: Into<Option<SockProtocol>>>(
     unsafe { Ok((OwnedFd::from_raw_fd(fds[0]), OwnedFd::from_raw_fd(fds[1]))) }
 }
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct Backlog(i32);
+
+impl Backlog {
+    /// Sets the listen queue size to system `SOMAXCONN` value
+    #[cfg(target_os = "redox")]
+    pub const DEFAULT: Self = Self(128);
+    #[cfg(not(target_os = "redox"))]
+    pub const DEFAULT: Self = Self(libc::SOMAXCONN);
+    /// Sets the listen queue size to -1 for system supporting it
+    #[cfg(any(target_os = "linux", target_os = "freebsd"))]
+    pub const MAXALLOWABLE: Self = Self(-1);
+    /// Reasonable upper limit, while in theory we might be able
+    /// to set it to a higher boundary, comes with a risk
+    /// of resource exhaustion in return
+    pub const MAXVALUE: Self = Self(511);
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum BacklogTryFromError {
+    TooNegative,
+    TooPositive,
+}
+
+impl std::fmt::Display for BacklogTryFromError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::TooNegative => write!(f, "Passed a positive backlog less than -1"),
+            Self::TooPositive => write!(f, "Passed a positive backlog greater than `511`")
+        }
+    }
+}
+
+impl std::error::Error for BacklogTryFromError {}
+
+impl<T: Into<Backlog>> From<Option<T>> for Backlog {
+    fn from(backlog: Option<T>) -> Self {
+        backlog.map_or(Self::DEFAULT, |b| b.into())
+    }
+}
+
+impl From<u16> for Backlog {
+    fn from(backlog: u16) -> Self {
+        Self(i32::from(backlog))
+    }
+}
+
+impl From<u8> for Backlog {
+    fn from(backlog: u8) -> Self {
+        Self(i32::from(backlog))
+    }
+}
+
+impl From<Backlog> for i32 {
+    fn from(backlog: Backlog) -> Self {
+        backlog.0
+    }
+}
+
+impl TryFrom<i64> for Backlog {
+    type Error = BacklogTryFromError;
+    fn try_from(backlog: i64) -> std::result::Result<Self, Self::Error> {
+        match backlog {
+            ..=-2 => Err(BacklogTryFromError::TooNegative),
+            -1..=511 => Ok(Self(i32::try_from(backlog).map_err(|_| BacklogTryFromError::TooPositive)?)),
```

Comment:
If a backlog of value `-1` is only valid on Linux and FreeBSD, why are we allowing it for other OSes here?

Reply:
sure. still a draft :) just wanted first impressions.